cpp/include/cudf/utilities/traits.hpp

```diff
@@ -370,5 +370,63 @@ constexpr inline bool is_nested(data_type type)
   return cudf::type_dispatcher(type, is_nested_impl{});
 }
 
+template <typename FromType, typename ToType>
+struct is_logically_castable_impl : std::false_type {
+};
+
+// Allow cast to same type
+template <typename Type>
+struct is_logically_castable_impl<Type, Type> : std::true_type {
+};
+
+#ifndef MAP_CASTABLE_TYPES
+#define MAP_CASTABLE_TYPES(Type1, Type2)                             \
+  template <>                                                        \
+  struct is_logically_castable_impl<Type1, Type2> : std::true_type { \
+  };                                                                 \
+  template <>                                                        \
+  struct is_logically_castable_impl<Type2, Type1> : std::true_type { \
+  };
+#endif
+
+// Allow cast from timestamp to integer representation
+MAP_CASTABLE_TYPES(cudf::timestamp_D, cudf::timestamp_D::duration::rep);
+MAP_CASTABLE_TYPES(cudf::timestamp_s, cudf::timestamp_s::duration::rep);
+MAP_CASTABLE_TYPES(cudf::timestamp_ms, cudf::timestamp_ms::duration::rep);
+MAP_CASTABLE_TYPES(cudf::timestamp_us, cudf::timestamp_us::duration::rep);
+MAP_CASTABLE_TYPES(cudf::timestamp_ns, cudf::timestamp_ns::duration::rep);
+
+template <typename FromType>
+struct is_logically_castable_to_impl {
+  template <typename ToType>
+  constexpr bool operator()()
+  {
+    return is_logically_castable_impl<FromType, ToType>::value;
+  }
+};
+
+struct is_logically_castable_from_impl {
+  template <typename FromType>
+  constexpr bool operator()(data_type to)
+  {
+    return type_dispatcher(to, is_logically_castable_to_impl<FromType>{});
+  }
+};
+
+/**
+ * @brief Indicates whether `from` is logically castable to `to`
+ *
+ * Column views that have binary compatible representations (e.g. timestamps and their integer
+ * representation) can be logically cast to another type.
+ *
+ * @param from The `data_type` to convert from
+ * @param to The `data_type` to convert to
+ * @return true if the types are logically castable
+ */
+constexpr bool is_logically_castable(data_type from, data_type to)
+{
+  return type_dispatcher(from, is_logically_castable_from_impl{}, to);
```

Comment:
I'm wondering how this should work for `fixed_point` types where we introduce metadata for the scale. Should such columns be allowed to cast to types with different scales, even though that would effectively corrupt the values?

Reply:
Changing the scale effectively just moves the decimal point, right? I'm not sure that would corrupt the values, but I don't think it's appropriate as a "cast" operation.

Reply:
I suppose this could actually be a useful and cheap operation if shifting the decimal point is precisely what you want to do, like converting from a percentage to a decimal. I don't think we can prevent casting from `fixed_point` to `int` and back to a `fixed_point` with a different scale in any case, so maybe the `fixed_point` to `fixed_point` case is ok.

Reply:
It does "corrupt" the value. If you have:
```cpp
auto x = decimal32{1.23, scale_type{-2}}; // Rep _value = 123 (10^2 * 1.23)
auto y = decimal32{1.23, scale_type{-3}}; // Rep _value = 1230 (10^3 * 1.23)
```
As you can see the, the internally stored `value` is different between the two, `123` vs `1230`.

Reply:
So this requires a regular (unary op) cast which knows how to rescale, right? In that case, it should be illegal to logically cast decimals with different scales?

Reply:
Right, the regular cast would be the way to rescale correctly, while the logical cast would just shift the decimal point producing a different value. However, moving the decimal point in this way could be potentially useful in any case where you want to multiply by a power of 10 for free (converting between SI units, % to decimal, etc). But if that were deemed to be useful, maybe it would be better to expose that more explicitly with a shift method or something.